packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts

```diff
@@ -0,0 +1,273 @@
+}
+
+enum SessionMode {
+  NoSession,
+  Private,
+  AuthenticatedPublic,
+}
+
+function sha(s: string) {
+  return createHash('sha256')
+    .update(s)
+    .digest('hex');
+}
+
+interface BaseCacheKey {
+  document: string;
+  operationName: string | null;
+  variables: { [name: string]: any };
+  extra: any;
+}
+
+interface ContextualCacheKey {
+  sessionMode: SessionMode;
+  sessionId?: string | null;
+}
+
+type CacheKey = BaseCacheKey & ContextualCacheKey;
+
+function cacheKeyString(key: CacheKey) {
+  return sha(JSON.stringify(key));
+}
+
+export default function plugin(
+  options: Options = Object.create(null),
+): ApolloServerPlugin {
+  return {
+    requestDidStart(
+      outerRequestContext: GraphQLRequestContext<any>,
+    ): GraphQLRequestListener<any> {
+      const cache = new PrefixingKeyValueCache(
+        options.cache || outerRequestContext.cache!,
+        'fqc:',
+      );
+
+      let sessionId: string | null = null;
+      let baseCacheKey: BaseCacheKey | null = null;
+
+      async function cacheGet(
+        contextualCacheKeyFields: ContextualCacheKey,
+      ): Promise<GraphQLResponse | null> {
+        const key = cacheKeyString({
+          ...baseCacheKey!,
+          ...contextualCacheKeyFields,
+        });
+        const value = await cache.get(key);
+        if (value === undefined) {
+          return null;
+        }
+        return JSON.parse(value);
+      }
+
+      function cacheSetInBackground(
+        contextualCacheKeyFields: ContextualCacheKey,
+        response: GraphQLResponse,
+        ttlSeconds: number,
+      ) {
+        const key = cacheKeyString({
+          ...baseCacheKey!,
+          ...contextualCacheKeyFields,
+        });
+        const value = JSON.stringify(response);
+        // Note that this function converts key and response to strings before
+        // doing anything asynchronous, so it can run in parallel with user code
+        // without worrying about anything being mutated out from under it.
+        cache.set(key, value, { ttl: ttlSeconds }).catch(console.warn);
+      }
+
+      return {
+        async execute(
+          requestContext: WithRequired<
+            GraphQLRequestContext<any>,
+            'document' | 'operationName' | 'operation'
+          >,
+        ): Promise<GraphQLResponse | null> {
+          // Call hooks. Save values which will be used in XXX as well.
+          let extraCacheKeyData: any = null;
+          if (options.sessionId) {
+            sessionId = await options.sessionId(requestContext);
+          }
+          if (options.extraCacheKeyData) {
+            extraCacheKeyData = await options.extraCacheKeyData(requestContext);
+          }
+
+          baseCacheKey = {
+            // XXX could also have requestPipeline add the unparsed document to requestContext;
+            // can't just use requestContext.request.query because that won't be set for APQs
+            document: print(requestContext.document),
```

Comment:
I'd really like to avoid having to print the document AST every time we need to compute a cache key. So the suggestion of putting the document text on the request context makes sense to me.

Reply:
Alternatively, we could replace the use of `document` and `operationName` by an `operationID`, which would rely on the same normalization we use for the operation registry. So we could add an `operationID` to the request context and use that across plugins.

Reply:
Gosh, the op reg stuff is so in flux though. Are you imagining this ID as just the pair of normalized doc and ID, or as an already hashed thing? I'm sort of not psyched about 100% standardizing on a cross-codebase normalization until  https://github.com/graphql/graphql-js/pull/1628 is merged. Though I also think it should be kosher to change cache key calculation later if we need. (Down for adding the string to the context though.)

Reply:
I was imagining the operation ID as a hash based on the normalized operation node + dependent fragments (so not the entire document, in case that contains multiple operations). I agree we want to use `stripIgnoredTokens` there. Realistically though, we can't expect everyone to be on the latest `graphql` even after that PR lands. Since it's only [a small bit of code](
https://github.com/graphql/graphql-js/pull/1628/files#diff-fea7ac4cd268610ae0fa84205423b0a7), maybe we should just include it for now?

Reply:
That would be my plan, but I'd like to see it merged and ideally released first, and the current PR has a comment about how the implementation may change.